packages/ckeditor5-engine/src/model/utils/insertcontent.js

```diff
@@ -380,113 +438,141 @@ class Insertion {
+		this.position = livePosition.toPosition();
+		livePosition.detach();
 
-			// OK:  <p>xx[]</p> + <p>yy</p> => <p>xx[]yy</p> (when sticks to previous)
-			// NOK: <p>xx[]</p> + <p>yy</p> => <p>xxyy[]</p> (when sticks to next)
-			const livePosition = LivePosition.fromPosition( this.position, 'toPrevious' );
+		// After merge elements that were marked by _insert() to be filtered might be gone so
+		// we need to mark the new container.
+		this._filterAttributesOf.push( this.position.parent );
 
-			// See comment above on moving `_affectedStart`.
-			if ( this._affectedEnd.isEqual( mergePosRight ) ) {
-				this._affectedEnd.detach();
-				this._affectedEnd = LivePosition._createAt( mergePosRight.nodeBefore, 'end', 'toNext' );
-			}
+		mergePosLeft.detach();
+	}
 
-			this.writer.merge( mergePosRight );
+	/**
+	 * Merges sibling of the last node if should be merged.
+	 *
+	 * @private
+	 */
+	_mergeSiblingsOfLastNode() {
+		const node = this._lastNode;
 
-			// See comment above on moving `_affectedStart`.
-			if ( mergePosRight.getShiftedBy( -1 ).isEqual( this._affectedStart ) && context.isFirst ) {
-				this._affectedStart.detach();
-				this._affectedStart = LivePosition._createAt( mergePosRight.nodeBefore, 0, 'toPrevious' );
-			}
+		if ( !( node instanceof Element ) ) {
+			return;
+		}
 
-			this.position = livePosition.toPosition();
-			livePosition.detach();
+		if ( !this._canMergeRight( node ) ) {
+			return;
 		}
 
-		if ( mergeLeft || mergeRight ) {
-			// After merge elements that were marked by _insert() to be filtered might be gone so
-			// we need to mark the new container.
-			this._filterAttributesOf.push( this.position.parent );
+		const mergePosRight = LivePosition._createAfter( node );
+		mergePosRight.stickiness = 'toNext';
+
+		/* istanbul ignore if */
+		if ( !this.position.isEqual( mergePosRight ) ) {
+			// Algorithm's correctness check. We should never end up here but it's good to know that we did.
+			// At this point the insertion position should be after the node we'll merge. If it isn't,
+			// it should need to be secured as in the left merge case.
+			/**
+			 * An internal error occurred when merging inserted content with its siblings.
+			 * The insertion position should equal the merge position.
+			 *
+			 * If you encountered this error, report it back to the CKEditor 5 team
+			 * with as many details as possible regarding the content being inserted and the insertion position.
+			 *
+			 * @error insertcontent-invalid-insertion-position
+			 */
+			throw new CKEditorError( 'insertcontent-invalid-insertion-position', this );
 		}
 
-		mergePosLeft.detach();
+		// Move the position to the previous node, so it isn't moved to the graveyard on merge.
+		// <p>x</p>[]<p>y</p> => <p>x[]</p><p>y</p>
+		this.position = Position._createAt( mergePosRight.nodeBefore, 'end' );
+
+		// OK:  <p>xx[]</p> + <p>yy</p> => <p>xx[]yy</p> (when sticks to previous)
```

Comment:
I know this is old comment, but I'd add this before it, because I didn't know what this example talks about:

```
// Explanation of setting position stickiness to `'toPrevious'`:
```